SELinux-Troubleshooting

dperpeet opened the pull request:
- [x] basic implementation using react
- [x] link to fixit (perform fix)
- [ ] ~~add option to view only alerts after a certain point in time~~
- [ ] ~~workaround: permanently dismiss alerts by adding journal entries~~
- [x] fix style
- [x] add jsx transformation to build step or commit js instead (significantly more difficult to read)
- [x] get integration tests to work

dperpeet commented:
for jsx you can download react-tools (e.g. `npm -g install react-tools`) and then transform the script via `cd pkg/setroubleshoot && jsx -x jsx views views`

stefwalter commented:
BTW, IMO this the package should be called 'selinux' with the troubleshooting stuff being one page, and related javascript files, CSS etc.

But as far as react, there's quite a bit to discuss about JSX. In particular in browser JSX transformation (while it works) is being shunned upstream, and also doesn't work with Content-Security-Policy. 

There are some ugly ways around this:
 * Abandon the concept of simple editing with just refreshing the browser (ie: no build step). Start requiring people to install Node.JS and run a build command before they can make a single change to Cockpit.
 * Following on from that ... Start using tools like Grunt during development to monitor files and skip the build command. This requires support in cockpit-bridge or similar.
 * Alternatively, Fork JSX transformer and hack our own version that works with CSP.

stefwalter commented:
Obviously another approach is to turn off Content-Security-Policy (or at least the unsafe-eval) option during development (which one can do with an override.json). However this makes it easy to ignore issues during development that would not pass CSP.

stefwalter commented:
JSXTransformer is *much* easier to switch to CSP compatible than babel is. That's because it uses a script insert, which we can setup with an nonce parameter.

dperpeet commented:
> Abandon the concept of simple editing with just refreshing the browser (ie: no build step). Start requiring people to install Node.JS and run a build command before they can make a single change to Cockpit.

I'd be fine with this. Actually, I only work this way - make and install to see changes in the browser.

stefwalter commented:
> I'd be fine with this. Actually, I only work this way - make and install to see changes in the browser.

Yes, but the point has been to lower the barrier to entry, not raise it. Requiring installing software to see changes is a big fail in 2016. There has to be a development mode that's easier to use.

dperpeet commented:
@andreasn This is a current screenshot with the 'if', 'then' and 'do' text for each possible solution:
![screenshot from 2016-03-11 15-04-30](https://cloud.githubusercontent.com/assets/8711649/13704288/64153460-e79b-11e5-87b7-819adb51657c.png)

andreasn commented:
Thanks!

andreasn commented:
"SELinux is preventing..." <- Shouldn't this be "SELinux prevented..." since it happened in the past?

dperpeet commented:
> "SELinux is preventing..." <- Shouldn't this be "SELinux prevented..." since it happened in the past?

This text is the summary of the alert as passed to us by setroubleshootd: https://github.com/bachradsusi/setroubleshoot/wiki/DBUS#return-values-3

If we gather a few of these issues we can open a pull request upstream and discuss.

dperpeet commented:
@andreasn current screenshot
![screenshot from 2016-03-17 09-21-33](https://cloud.githubusercontent.com/assets/8711649/13840292/ebe5645c-ec21-11e5-8b2c-1ceb778e05c8.png)

andreasn commented:
"4 occurrences" should go under the delete icon in order to match up better with how these panels are used elsewhere (OSTree, Kubernetes, Registry)

andreasn commented:
I wonder if a more correct label for "Action:" would be "Solution". That would map better with "Apply this solution"

dperpeet commented:
bot: Image refresh for fedora-24

cockpituous commented:
Image creation for fedora-24 in process on host32-rack06.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-fedora-24-7160e2f7d858023e13fd357488032debeefa0568/

cockpituous commented:
Image creation failed.

dperpeet commented:
bot: Image refresh for fedora-24

cockpituous commented:
Image creation for fedora-24 in process on host37-rack09.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-fedora-24-4ae736297f7592eed9226caaba4da450f9cd480a/

cockpituous commented:
Image creation failed.

dperpeet commented:
addressed issues, updated screenshot:
![screenshot from 2016-03-18 15-23-09](https://cloud.githubusercontent.com/assets/8711649/13880667/a0db2922-ed1d-11e5-92bc-f4a5344537cc.png)

stefwalter commented on pkg/base1/cockpit.js:
```diff
@@ -43,6 +43,19 @@ function is_array(it) {
     return Object.prototype.toString.call(it) === '[object Array]';
 }
 
+// A shim for non ES5 supporting browsers, like PhantomJS. Lovingly inspired by:
```
We decided a long time ago we weren't going to modify builtin objects and polyfill them from the base1 package. So I guess the question is, could we make this code just run on phantomjs?

stefwalter replied:
Also an explanation of what this code is expected by would probably be good.

stefwalter replied:
This still needs to be removed discussed or adjusted.

andreasn commented:
UI looks good to me now!
And, as a bonus, I was able to fix a SELinux error on one of my VMs!

stefwalter commented on pkg/base1/Makefile.am:
```diff
@@ -16,6 +16,8 @@ nodist_base_DATA = \
 	pkg/base1/cockpit.min.js.gz \
 	pkg/base1/bundle.min.js.gz \
+	pkg/base1/react.min.js.gz \
+	pkg/base1/react-dom.min.js.gz \
```
The split is an annoying react choice. But will do a follow up replacing this with react-lite, so not an issue.

stefwalter commented on pkg/base1/Makefile.am:
```diff
@@ -131,6 +137,8 @@ CLEANFILES += \
 	pkg/base1/moment.min.js \
 	pkg/base1/mustache.min.js \
+  pkg/base1/react.min.js \
+  pkg/base1/react-dom.min.js \
```
Indentation fix needed.

stefwalter commented on pkg/selinux/Makefile.am:
```diff
@@ -0,0 +1,73 @@
+setroubleshoot_MODULES = \
+	$(PKGDIR)/setroubleshoot.js \
+  $(PKGDIR)/setroubleshoot-client.js \
+  $(PKGDIR)/views/setroubleshoot-view.js \
```
Indentation.

stefwalter commented on pkg/selinux/Makefile.am:
```diff
@@ -0,0 +1,73 @@
+	$(NULL)
+setroubleshoot_DATA = \
+ 	$(PKGDIR)/manifest.json \
+  $(PKGDIR)/setroubleshoot-app.js \
```
indentation.

stefwalter commented on pkg/selinux/Makefile.am:
```diff
@@ -0,0 +1,73 @@
+	$(PKGDIR)/setroubleshoot.min.css \
+	$(PKGDIR)/setroubleshoot.min.html \
+	$(PKGDIR)/bundle.min.js \
+  $(PKGDIR)/setroubleshoot-app.min.js \
```
indentation

stefwalter commented on pkg/selinux/manifest.json:
```diff
@@ -0,0 +1,11 @@
+            "order": 8
+        }
+    },
+    "content-security-policy": "default-src 'self'"
```
This line isn't necessary.

dperpeet replied:
it broke for me without that line, I'll test again

dperpeet replied:
> it broke for me without that line, I'll test again

it doesn't work for me without the line

stefwalter commented on pkg/selinux/setroubleshoot-app.js:
```diff
@@ -0,0 +1,7 @@
+function(troubleshoot) {
+    const app = document.getElementById('app');
+    troubleshoot.init(app);
+});
```
Seems like this file should be merged with setroubleshoot.js ... that's what I've been doing with all the other places where we're moving code out of the HTML due to CSP.

dperpeet replied:
> Seems like this file should be merged with setroubleshoot.js ... that's what I've been doing with all the other places where we're moving code out of the HTML due to CSP.

On the other hand, this way we have cleaner separation. I noticed I haven't removed the jquery dependency in setroubleshoot.js, but I could. That way it could be included from anywhere, with any top level html element.
I guess the question is whether to keep this glue between the actual component and the rendering target or merge them like you did before. I like the separation, but for the sake of having fewer files I could go with the merged variant.

andreasn commented:
The way "solution details" jumps a bit when clicked is slightly jarring.
Setting a width of 8px on the i-tag with fa seems to help.

andreasn commented:
It breaks the expansion behavior that the green banner shows up in the middle between the expander and the block it's expanding.

![screenshot from 2016-03-24 11-50-43](https://cloud.githubusercontent.com/assets/209895/14015062/aad3d242-f1b6-11e5-88db-059901b25a68.png)

dperpeet commented:
> It breaks the expansion behavior that the green banner shows up in the middle between the expander and the block it's expanding.

what do you suggest? beneath the details could mean that the result is hidden (because the details can be very, very long)

Moving it up means that the green box can collide with the button.

andreasn commented:
> what do you suggest? beneath the details could mean that the result is hidden (because the details can be very, very long)

But only if the details are expanded, right?
I think that's ok behavior in that case.

dperpeet commented:
> But only if the details are expanded, right?
> I think that's ok behavior in that case.

Wouldn't standard behavior be to look at the details and then press "apply"? In that case I wouldn't see the result of the operation without scrolling. I can also auto-collapse when the user hits apply, that would solve the issue.

andreasn commented:
> I can also auto-collapse when the user hits apply, that would solve the issue.

That sounds like a neat solution!

dperpeet commented:
>> I can also auto-collapse when the user hits apply, that would solve the issue.
> 
> That sounds like a neat solution!

This is how it works now.

dperpeet commented:
issues addressed, tests passed on fedora-24
